cpp/src/io/orc/orc.cpp

```diff
@@ -402,6 +360,181 @@ const uint8_t *OrcDecompressor::Decompress(const uint8_t *srcBytes, size_t srcLe
+                                                              size_type &row_count)
+{
+  std::vector<OrcStripeInfo> selection;
+
+  if (!stripes.empty()) {
+    size_t stripe_rows = 0;
+    for (const auto &stripe_idx : stripes) {
+      CUDF_EXPECTS(stripe_idx >= 0 && stripe_idx < get_num_stripes(), "Invalid stripe index");
+      selection.emplace_back(&ff.stripes[stripe_idx], nullptr);
+      stripe_rows += ff.stripes[stripe_idx].numberOfRows;
+    }
+    // row_start is 0 if stripes are set. If this is not true anymore, then
+    // row_start needs to be subtracted to get the correct row_count
+    CUDF_EXPECTS(row_start == 0, "Start row index should be 0");
+    row_count = static_cast<size_type>(stripe_rows);
+  } else {
+    row_start = std::max(row_start, 0);
+    if (row_count < 0) {
+      row_count = static_cast<size_type>(
+        std::min<size_t>(get_total_rows() - row_start, std::numeric_limits<size_type>::max()));
+    } else {
+      row_count = static_cast<size_type>(std::min<size_t>(get_total_rows() - row_start, row_count));
+    }
+    CUDF_EXPECTS(row_count >= 0 && row_start >= 0, "Negative row count or starting row");
+    CUDF_EXPECTS(
+      !(row_start > 0 && (row_count > (std::numeric_limits<size_type>::max() - row_start))),
+      "Summation of starting row index and number of rows would cause overflow");
+
+    size_type stripe_skip_rows = 0;
+    for (size_t i = 0, count = 0; i < ff.stripes.size(); ++i) {
+      count += ff.stripes[i].numberOfRows;
+      if (count > static_cast<size_t>(row_start)) {
+        if (selection.empty()) {
+          stripe_skip_rows =
+            static_cast<size_type>(row_start - (count - ff.stripes[i].numberOfRows));
+        }
+        selection.emplace_back(&ff.stripes[i], nullptr);
+      }
+      if (count >= static_cast<size_t>(row_start) + static_cast<size_t>(row_count)) { break; }
+    }
+    row_start = stripe_skip_rows;
+  }
+
+  // Read each stripe's stripefooter metadata
+  if (not selection.empty()) {
+    stripefooters.resize(selection.size());
+    for (size_t i = 0; i < selection.size(); ++i) {
+      const auto stripe         = selection[i].first;
+      const auto sf_comp_offset = stripe->offset + stripe->indexLength + stripe->dataLength;
+      const auto sf_comp_length = stripe->footerLength;
+      CUDF_EXPECTS(sf_comp_offset + sf_comp_length < source->size(), "Invalid stripe information");
+
+      const auto buffer = source->host_read(sf_comp_offset, sf_comp_length);
+      size_t sf_length  = 0;
+      auto sf_data      = decompressor->Decompress(buffer->data(), sf_comp_length, &sf_length);
+      ProtobufReader(sf_data, sf_length).read(stripefooters[i]);
+      selection[i].second = &stripefooters[i];
+    }
+  }
+
+  return selection;
+}
+
+std::vector<int> metadata::select_columns(std::vector<std::string> use_names,
+                                          bool &has_timestamp_column)
```

Comment:
Output parameters are a code smell for me. Happens in a couple places. Not sure if possible to easily avoid.

Reply:
I completely agree. We have an [issue](https://github.com/rapidsai/cudf/issues/5943) filed for this. Is it okay to address this as a part of that issue?
IMO there should be a separate function that determines if there are timestamp columns, instead of making this function do two things to save a few CPU cycles.